File: src/at/greywind/onlinereader/DBManager.java
package at.greywind.onlinereader;

import javax.servlet.ServletContext;
import java.sql.*;
import java.util.ArrayList;

public class DBManager {
    private Connection connect = null;

    private static String dbName = null;
    private static String dbPassword = null;
    private static String dbUser = null;

    public static void main(String[] args){
        DBManager m = new DBManager();
        try {
            m.deleteBookFromUser(75, 5);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public static boolean isDatabaseSet(){
        if(dbName != null && dbPassword != null && dbUser != null)
            return true;

        return false;
    }

    public static void setDatabase(ServletContext context){
        dbName = context.getInitParameter("db-name");
        dbPassword = context.getInitParameter("db-password");
        dbUser = context.getInitParameter("db-user");
    }

    public DBManager(){
        try {
            Class.forName("com.mysql.jdbc.Driver");
            connect = DriverManager.getConnection("jdbc:mysql://localhost/"+ dbName + "?" + "user="+ dbUser + "&password=" + dbPassword);
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        } catch (SQLException e) {
            e.printStackTrace();
        }

    }

    public User register(String username, String email, String password, String confirmPassword) throws NoSuchUserNameException, WrongPasswordException, SQLException {
        Statement statement = null;
        ResultSet resultSet = null;
        try {
            statement = connect.createStatement();
            resultSet = statement.executeQuery("SELECT * FROM user WHERE username = '" + username.toLowerCase() + "'");

            if (resultSet.isBeforeFirst()) {
                throw new NoSuchUserNameException();
            }

            closeSetAndStatement(resultSet, statement);

            statement = connect.createStatement();
            resultSet = statement.executeQuery("SELECT * FROM user WHERE email = '" + email.toLowerCase() + "'");

            if (resultSet.isBeforeFirst()) {
                throw new EmailAlreadyExistsException();
            }

            closeSetAndStatement(resultSet, statement);

            if(!password.equals(confirmPassword)){
                throw new WrongPasswordException();
            }

            PreparedStatement preparedStatement = connect.prepareStatement("INSERT INTO online_reader.user VALUE (default, ?, ?, ?)");
            preparedStatement.setString(1, username.toLowerCase());
            preparedStatement.setString(2, email.toLowerCase());
            preparedStatement.setString(3, password);
            preparedStatement.executeUpdate();
            preparedStatement.close();

            statement = connect.createStatement();
            resultSet = statement.executeQuery("SELECT * FROM user WHERE email = '" + email.toLowerCase() + "'");
            resultSet.first();

            return new User(resultSet.getInt("id"), resultSet.getString("username"),resultSet.getString("email"),resultSet.getString("password"));
        }catch (Exception e){
            throw e;
        }finally {
            closeSetAndStatement(resultSet, statement);
        }
    }

    public User login(String username, String password) throws WrongPasswordException, NoSuchUserNameException, SQLException {
        ResultSet resultSet = null;
        Statement statement = null;
        try {
            statement = connect.createStatement();
            resultSet = statement.executeQuery("SELECT * FROM user WHERE username = '" + username + "'");

            if (!resultSet.isBeforeFirst() ) {
               throw new NoSuchUserNameException();
            }

            resultSet.first();
            if(!password.equals(resultSet.getString("password"))) {
                throw new WrongPasswordException();
            }

            return new User(resultSet.getInt("id"), resultSet.getString("username"),resultSet.getString("email"),resultSet.getString("password"));
        } catch (Exception e) {
            throw e;
        } finally {
            closeSetAndStatement(resultSet, statement);
        }
    }

    public void deleteBookFromUser(int bookID, int userID) throws SQLException {
        Statement statement = connect.createStatement();
        statement.execute("DELETE FROM online_reader.user_has_book WHERE user_id='"+ userID +"' and book_id='" + bookID + "'");
        statement.close();

        statement = connect.createStatement();
        statement.execute("DELETE FROM online_reader.book WHERE id='" + bookID + "'");
        statement.close();
    }

    public void addBookToUser(String title, String description, String filename, int userID) throws SQLException{
        Statement statement = null;
        ResultSet resultSet = null;
        try {
            PreparedStatement preparedStatement = connect.prepareStatement("INSERT INTO online_reader.book VALUE (default, ?, ?, ?)");
            preparedStatement.setString(1, filename);
            preparedStatement.setString(2, description);
            preparedStatement.setString(3, title);
            preparedStatement.executeUpdate();
            preparedStatement.close();

            statement = connect.createStatement();
            resultSet = statement.executeQuery("SELECT * FROM book WHERE book.name = '" + filename + "'");
            resultSet.first();

            int bookID = resultSet.getInt("id");

            preparedStatement = connect.prepareStatement("INSERT INTO online_reader.user_has_book VALUE (?, ?)");
            preparedStatement.setInt(1, userID);
            preparedStatement.setInt(2, bookID);
            preparedStatement.executeUpdate();
            preparedStatement.close();
        }catch (Exception e){
               throw e;
        }finally {
            closeSetAndStatement(resultSet, statement);
        }
    }

    public ArrayList<Book> getBooksforUser(int userID){
        ArrayList<Book> books = new ArrayList<>();
        Statement statement = null;
        try {
            statement = connect.createStatement();
            ResultSet resultSet = statement.executeQuery("select book.id, book.name, book.description, book.title from book inner join user_has_book on book.id = user_has_book.book_id where user_has_book.user_id = " + userID +";");

            while (resultSet.next()){
                books.add(new Book(resultSet.getInt("id"), resultSet.getString("name"), resultSet.getString("title"), resultSet.getString("description")));
            }

        } catch (SQLException e) {
            e.printStackTrace();
        }

        return books;
    }

    private void closeSetAndStatement(ResultSet resultSet, Statement statement) {
        if(resultSet != null) {
            try {
                resultSet.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }else if(statement != null){
            try {
                statement.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
    }

    public void close() {
        try {
            if (connect != null) {
                connect.close();
            }
        } catch (Exception e) {

        }
    }
}

File: web/src/DBManager.php
<?php
/**
 * Created by PhpStorm.
 * User: quentinwendegass
 * Date: 18.04.18
 * Time: 13:01
 */

require_once "Exceptions.php";
require_once "Book.php";
require_once "User.php";
require_once "Constants.php";

class DBManager
{

    public function __construct()
    {

    }

    /**
     * @param $username
     * @param $email
     * @param $password
     * @param $confirmPassword
     * @return User
     * @throws EmailAlreadyExistsException
     * @throws UserAlreadyExistsException
     * @throws WrongPasswordException
     * @throws SQLException
     */
    public function register($username, $email, $password, $confirmPassword)
    {
        $conn = new mysqli(DB_HOST, DB_USER, DB_PASSWORD, DB_NAME);
        if ($conn->connect_error) {
            throw new SQLException();
        }

        $sql = "SELECT * FROM user WHERE username = ?";
        $stmt = $conn->prepare($sql);
        $stmt->bind_param("s", strtolower($username));

        $stmt->execute();
        $result = $stmt->get_result();

        if ($result->num_rows > 0){
            $result->close();
            $stmt->close();
            $conn->close();
            throw new UserAlreadyExistsException();
        }

        $result->close();
        $stmt->close();

        $sql = "SELECT * FROM user WHERE email = ?";
        $stmt = $conn->prepare($sql);
        $stmt->bind_param("s", strtolower($email));

        $stmt->execute();
        $result = $stmt->get_result();

        if ($result->num_rows > 0)
            throw new EmailAlreadyExistsException();

        $result->close();

        if($password != $confirmPassword){
            $conn->close();
            throw new WrongPasswordException();
        }

        $sql = "INSERT INTO online_reader.user VALUES (default, ?, ?, ?)";
        $stmt = $conn->prepare($sql);
        $stmt->bind_param("sss", $username, $email, $password);

        $stmt->execute();
        $stmt->close();

        $sql = "SELECT * FROM user WHERE email = ?";
        $stmt = $conn->prepare($sql);
        $stmt->bind_param("s",strtolower($email));

        $stmt->execute();

        $result = $stmt->get_result();

        $userData = $result->fetch_assoc();

        $user = new User($userData["id"], $userData["username"], $userData["email"], $userData["password"]);

        $result->close();
        $stmt->close();
        $conn->close();

        return $user;
    }

    /**
     * @param $username
     * @param $password
     * @return User
     * @throws NoSuchUserNameException
     * @throws SQLException
     * @throws WrongPasswordException
     */
    public function login($username, $password)
    {
        $conn = new mysqli(DB_HOST, DB_USER, DB_PASSWORD, DB_NAME);
        if ($conn->connect_error) {
            throw new SQLException();
        }

        $sql = "SELECT * FROM user WHERE username = ?";
        $stmt = $conn->prepare($sql);
        $stmt->bind_param("s",strtolower($username));

        $stmt->execute();
        $result = $stmt->get_result();

        if ($result->num_rows == 0){
            $result->close();
            $stmt->close();
            $conn->close();
            throw new NoSuchUserNameException();
        }

        $userData = $result->fetch_assoc();

        if($userData["password"] != $password){
            $result->close();
            $stmt->close();
            $conn->close();
            throw new WrongPasswordException();
        }

        $user = new User($userData["id"], $userData["username"], $userData["email"], $userData["password"]);

        $result->close();
        $stmt->close();
        $conn->close();

        return $user;
    }

    /**
     * @param $bookID
     * @param $userID
     * @throws SQLException
     */
    public function deleteBookFromUser($bookID, $userID)
    {
        $conn = new mysqli(DB_HOST, DB_USER, DB_PASSWORD, DB_NAME);
        if ($conn->connect_error) {
            throw new SQLException();
        }

        $sql = "DELETE FROM online_reader.user_has_book WHERE user_id= ? and book_id = ?";

        $stmt = $conn->prepare($sql);
        $stmt->bind_param("ii",$userID, $bookID);


        if ($stmt->execute() === FALSE) {
            $stmt->close();
            $conn->close();
            throw new SQLException();
        }

        $stmt->close();

        $sql = "DELETE FROM online_reader.book WHERE id = ?";

        $stmt = $conn->prepare($sql);
        $stmt->bind_param("i",$bookID);

        if ($stmt->execute() === FALSE) {
            $stmt->close();
            $conn->close();
            throw new SQLException();
        }

        $stmt->close();
        $conn->close();
    }

    /**
     * @param $title
     * @param $description
     * @param $filename
     * @param $userID
     * @throws NoSuchElementException
     * @throws SQLException
     */
    public function addBookToUser($title, $description, $filename, $userID)
    {
        $conn = new mysqli(DB_HOST, DB_USER, DB_PASSWORD, DB_NAME);
        if ($conn->connect_error) {
            throw new SQLException();
        }

        $sql = "INSERT INTO online_reader.book VALUES (default, ?, ?, ?)";
        $stmt = $conn->prepare($sql);
        $stmt->bind_param("sss", $filename, $description, $title);

        $stmt->execute();
        $stmt->close();

        $sql = "SELECT * FROM book WHERE book.name = ?";
        $stmt = $conn->prepare($sql);
        $stmt->bind_param("s",$filename);

        $stmt->execute();
        $result = $stmt->get_result();

        if ($result->num_rows == 0){
            $stmt->close();
            $result->close();
            $conn->close();
            throw new NoSuchElementException();
        }

        $bookData = $result->fetch_assoc();
        $bookID = $bookData["id"];

        $result->close();
        $stmt->close();

        $sql = "INSERT INTO online_reader.user_has_book VALUES (?, ?)";
        $stmt = $conn->prepare($sql);
        $stmt->bind_param("ii", $userID, $bookID);

        $stmt->execute();
        $stmt->close();

        $conn->close();
    }

    /**
     * @param $userID
     * @return array
     * @throws NoSuchElementException
     * @throws SQLException
     */
    public function getBooksForUser($userID)
    {
        $books = array();

        $conn = new mysqli(DB_HOST, DB_USER, DB_PASSWORD, DB_NAME);
        if ($conn->connect_error) {
            throw new SQLException();
        }

        $sql = "select book.id, book.name, book.description, book.title from book inner join user_has_book on book.id = user_has_book.book_id where user_has_book.user_id = ?";
        $stmt = $conn->prepare($sql);
        $stmt->bind_param("i", $userID);

        $stmt->execute();
        $result = $stmt->get_result();

        if ($result->num_rows > 0){
            while($row = $result->fetch_assoc()) {
                $books[] = new Book($row["id"], $row["name"], $row["title"], $row["description"]);
            }
        }else{
            $result->close();
            $stmt->close();
            $conn->close();
            throw new NoSuchElementException();
        }

        $result->close();
        $stmt->close();
        $conn->close();

        return $books;
    }
}